Add bucket index support to querier

pracucci opened the pull request:
**What this PR does**:
In this PR I'm adding the bucket index support to querier. Using the bucket index in the querier is optional and disabled by default (and will be until marked experimental).

**Which issue(s) this PR fixes**:
N/A

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

pstibrany commented on pkg/storage/tsdb/caching_bucket.go:
```diff
@@ -92,6 +94,8 @@ func (cfg *MetadataCacheConfig) RegisterFlagsWithPrefix(f *flag.FlagSet, prefix
 	f.DurationVar(&cfg.MetafileContentTTL, prefix+"metafile-content-ttl", 24*time.Hour, "How long to cache content of the metafile.")
 	f.IntVar(&cfg.MetafileMaxSize, prefix+"metafile-max-size-bytes", 1*1024*1024, "Maximum size of metafile content to cache in bytes.")
 	f.DurationVar(&cfg.MetafileAttributesTTL, prefix+"metafile-attributes-ttl", 168*time.Hour, "How long to cache attributes of the block metafile.")
+	f.DurationVar(&cfg.BucketIndexContentTTL, prefix+"bucket-index-content-ttl", 5*time.Minute, "How long to cache content of the metafile.")
```
s/metafile/bucket index/

pstibrany commented on pkg/storage/tsdb/bucketindex/loader.go:
```diff
@@ -0,0 +1,253 @@
+	// Build a list of users for which we should update or delete the index.
+	l.indexesMx.RLock()
+	for userID, entry := range l.indexes {
+		if time.Since(entry.getRequestedAt()) >= l.cfg.IdleTimeout {
```
Nit: Candidate for `switch { ... }` to reduces some else-ifs.

pstibrany commented on pkg/storage/tsdb/bucketindex/loader.go:
```diff
@@ -0,0 +1,253 @@
+	// Delete unused indexes, if confirmed they're still unused.
+	l.indexesMx.Lock()
+	for _, userID := range toDelete {
+		if idx := l.indexes[userID]; time.Since(idx.getRequestedAt()) >= l.cfg.IdleTimeout {
```
nit: reuse `time.Now()` between iterations?

pstibrany commented on pkg/storage/tsdb/bucketindex/loader.go:
```diff
@@ -0,0 +1,253 @@
+
+	// Update actively used indexes.
+	for _, userID := range toUpdate {
+		readCtx, cancel := context.WithTimeout(ctx, readIndexTimeout)
```
To avoid linters panicking due to defer in the loop, we could move entire for body into a method. It would also make `defer` more efficient.

pstibrany commented on pkg/storage/tsdb/bucketindex/loader.go:
```diff
@@ -0,0 +1,253 @@
+	err   error
+
+	// Unix timestamp (seconds) of when the index has been updated from the storage the last time.
+	updatedAt *atomic.Int64
```
Nit: No need for pointers on atomic values, if outer structure (`cachedIndex`) is used via pointers.

pstibrany commented on pkg/storage/tsdb/bucketindex/loader.go:
```diff
@@ -0,0 +1,253 @@
+
+		// We don't check if the index is stale because it's the responsibility
+		// of the background job to keep it updated.
+		entry.requestedAt.Store(time.Now().Unix())
```
I'd suggest making update of requestedAt timestamp a method on `*cachedIndex`, to keep the logic at single place.

pstibrany commented on pkg/querier/blocks_finder_bucket_index_test.go:
```diff
@@ -0,0 +1,219 @@
+	require.NoError(b, bucketindex.WriteIndex(ctx, bkt, userID, idx))
+	finder := prepareBucketIndexBlocksFinder(b, bkt)
+
+	for n := 0; n < b.N; n++ {
```
`b.ResetTimer()` before running benchmark itself?

pstibrany commented on pkg/storage/tsdb/bucketindex/loader_test.go:
```diff
@@ -0,0 +1,426 @@
+		cortex_bucket_index_loaded 1
+	`),
+		"cortex_bucket_index_load_failures_total",
+		"cortex_bucket_index_loaded",
```
Should we verify `cortex_bucket_index_loads_total` as well? (I guess it could be bit flaky though)

pracucci replied:
I didn't in this (and few other) test because it would be flaky. There's no easy way to check it in a stable way, because we have no control over the background checks.

pstibrany reviewed (approved):
Wow, great job! No major problems found, logic is solid. :clap:

pracucci commented on pkg/querier/blocks_finder_bucket_index.go:
```diff
@@ -0,0 +1,137 @@
+		matchingBlocks[block.ID] = block
+	}
+
+	for _, mark := range idx.BlockDeletionMarks {
```
Yes, that was an idea. Then I wrote a benchmark (1K blocks, 100 marks) and it takes 40Kns for each `GetBlocks()` call, reason why I believe it's not worth it.

pracucci commented on pkg/querier/blocks_finder_bucket_index.go:
```diff
@@ -0,0 +1,137 @@
+		return nil, err
+	}
+
+	f.Service = services.NewBasicService(f.starting, f.running, f.stopping)
```
Very good point!

pracucci commented:
Thanks @pstibrany for your deep review! After reading your comments I realised it was definitely a draft what I've submitted 😉  I've addressed all comments, except a couple I've commented about.

pstibrany commented on docs/blocks-storage/bucket-index.md:
```diff
@@ -0,0 +1,42 @@
+
+The bucket index is a **per-tenant file containing the list of blocks and block deletion marks** in the storage. The bucket index itself is stored in the backend object storage, is periodically updated by the compactor and used by queriers to discover blocks in the storage.
+
+The bucket index usage is **optional** and can be enabled via `-blocks-storage.bucket-store.bucket-index.enabled=true` (or its respective YAML config option).
```
Explain why it is useful?

pracucci replied:
Done. WDYT?

pstibrany commented on docs/blocks-storage/bucket-index.md:
```diff
@@ -0,0 +1,42 @@
+
+If a bucket index is unused for a long time (configurable via `-blocks-storage.bucket-store.bucket-index.idle-timeout`), e.g. because that querier instance is not receiving any query from the tenant, the querier will offload it, stopping to keep it updated at regular intervals. This is particularly for tenants which are resharded to different queriers when [shuffle sharding](../guides/shuffle-sha ...
+
+Finally, the querier, at query time, checks how old is a bucket index (based on its `updated_at`) and fail a query if its age is older than `-blocks-storage.bucket-store.bucket-index.max-stale-period`. This circuit breaker is used to ensure queriers will not return any partial query results due to a stale view over the long-term storage.
```
You have also added caching of index into caching-bucket. Is that worth mentioning as well?

pracucci replied:
Yes, sure, done. WDYT?

pstibrany commented on docs/blocks-storage/compactor.md:
```diff
@@ -10,7 +10,7 @@ slug: compactor
 
 - Compact multiple blocks of a given tenant into a single optimized larger block. This helps to reduce storage costs (deduplication, index size reduction), and increase query speed (querying fewer blocks is faster).
-- Keep the per-tenant bucket index updated. The bucket index is used by [queriers](./querier.md) and [store-gateways](./store-gateway.md) to discover new blocks in the storage.
+- Keep the per-tenant bucket index updated. The [bucket index](./bucket-index.md) is used by [queriers](./querier.md) and [store-gateways](./store-gateway.md) to discover new blocks in the storage.
```
In the bucket-index.md only queriers are mentioned. Let's make it consistent (either way is fine, but store-gateways support is not yet there).

pracucci replied:
My plan is to update doc for store-gateway once support will be there. I will do it in the store-gateway PR.

pstibrany replied:
My point is that this document already mentions store-gateways, but bucket-index.md doesn't, which may be confusing. Either mention it at both places, or not at all?

pracucci replied:
Oh right. I removed the mention to store-gateway. I will update it in the store-gateway PR.

pstibrany commented on docs/blocks-storage/querier.md:
```diff
@@ -13,12 +13,28 @@ The querier is **stateless**.
 
+### Bucket index enabled
+
+When [bucket index](./bucket-index.md) is enabled, queriers lazily download the bucket index upon the 1st query received for a given tenant, cache it in memory and periodically keep it update. The bucket index contains the list of blocks and block deletion marks of a tenant, which is later used during the query execution to find the set of blocks that need to be queried for the given query.
```
s/1st/first/

ranton256 commented on pkg/storage/tsdb/bucketindex/loader.go:
```diff
@@ -0,0 +1,269 @@
+	)
+
+	// Build a list of users for which we should update or delete the index.
+	l.indexesMx.RLock()
```
I would suggest making this a helper method to calculate toDelete and toUpdate since it's the scope that needs the lock and doing something separate to make things slightly easier to read.

pracucci replied:
Good point, done. What do you think=

ranton256 replied:
Thanks, LGTM

pracucci commented on docs/blocks-storage/_index.md:
```diff
@@ -29,7 +29,7 @@ When running the Cortex blocks storage, the Cortex architecture doesn't signific
 The **[store-gateway](./store-gateway.md)** is responsible to query blocks and is used by the [querier](./querier.md) at query time. The store-gateway is required when running the blocks storage.
 
-The **[compactor](./compactor.md)** is responsible to merge and deduplicate smaller blocks into larger ones, in order to reduce the number of blocks stored in the long-term storage for a given tenant and query them more efficiently. It also keeps the bucket index updated and, for this reason, it's a required component.
+The **[compactor](./compactor.md)** is responsible to merge and deduplicate smaller blocks into larger ones, in order to reduce the number of blocks stored in the long-term storage for a given tenant and query them more efficiently. It also keeps the [bucket index](./bucket-index.md) updated and, for this reason, it's a required component.
```
> Is this going to cause issues for users running with compactor now(not sure if there are such people)?

Did you mean "with compactor" or "without compactor"? I'm no sure I understand the question.

> Also, bucket-index.md says the bucket index is optional which seems like it contradicts this statement.

The compactor always writes the bucket index. The flag to enable bucket index is whether it should be used or not in queriers (and in the store-gateway in the upcoming PR too). The point is that, before enabling the bucket index in the queriers and store-gateway, you have to rollout the compactor first, so that the bucket index for all tenants is created before you enable it in querier and store-gateway. To simplify it (at least I thought it would have simplified), I've kept it always enabled in the compactor.

pracucci commented on docs/blocks-storage/bucket-index.md:
```diff
@@ -0,0 +1,42 @@
+
+If a bucket index is unused for a long time (configurable via `-blocks-storage.bucket-store.bucket-index.idle-timeout`), e.g. because that querier instance is not receiving any query from the tenant, the querier will offload it, stopping to keep it updated at regular intervals. This is particularly for tenants which are resharded to different queriers when [shuffle sharding](../guides/shuffle-sha ...
+
+Finally, the querier, at query time, checks how old is a bucket index (based on its `updated_at`) and fail a query if its age is older than `-blocks-storage.bucket-store.bucket-index.max-stale-period`. This circuit breaker is used to ensure queriers will not return any partial query results due to a stale view over the long-term storage.
```
I don't think falling back is a viable solution. The "bucket scan" logic requires a preventive bucket scanning, which we don't do when we enable the bucket index. Lazily bucket scanning would be too slow to do at query time. Moreover, fallback logic unfrequently exercised would bring further risks that fallback logic doesn't work as expected.

In my opinion, when you run Cortex with bucket index, the bucket index is an essential part of the system and it's required to be kept updated. The compactor already exports a metric with the timestamp of the last time the bucket index of each tenant has been updated, so that we can alert on it before the `max-stale-period` is reached.

pracucci replied:
What do you think?

pracucci commented on docs/blocks-storage/querier.md:
```diff
@@ -587,6 +604,14 @@ blocks_storage:
+
+      # Maximum size of bucket index content to cache in bytes.
+      # CLI flag: -blocks-storage.bucket-store.metadata-cache.bucket-index-max-size-bytes
+      [bucket_index_max_size_bytes: <int> | default = 1048576]
```
Yes. I will update the doc.

To answer you, it doesn't get cached (skipped to be cached). Bucket index should never hit such big size tho. We see less 50KB compressed for 1K blocks.

pracucci replied:
Done. Does the new description clarifies it?

pracucci commented on docs/blocks-storage/bucket-index.md:
```diff
@@ -0,0 +1,42 @@
+
+The [compactor](./compactor.md) periodically scans the bucket and uploads an updated bucket index to the storage. The frequency at which the bucket index is updated can be configured via `-compactor.cleanup-interval`.
+
+The bucket index is built and updated by the compactor even if `-blocks-storage.bucket-store.bucket-index.enabled` has **not** been enabled. This is intentional and the overhead introduced by keeping the bucket index is non significative.
```
I've updated the doc. Is it more clear now?

pracucci commented:
@pstibrany @ranton256 I'm going to merge this PR, but feel free to add further comments (if any). I will promptly address post-merge comments too.
